tests/run/opaque-immutable-array-xm.scala

```diff
@@ -0,0 +1,51 @@
+import scala.reflect.ClassTag
+object Test extends App {
+
+  import java.util.Arrays
+
+  opaque type IArray[A1] = Array[A1]
+
+  implicit object IArray {
+    inline def initialize[A](body: => Array[A]): IArray[A] = body
+    def apply[A: ClassTag](xs: A*): IArray[A] = initialize(Array(xs: _*))
+
+    // These should be inline but that does not work currently. Try again
+    // once inliner is moved to ReifyQuotes
+    def (ia: IArray[A]) length[A]: Int = (ia: Array[A]).length
+    def (ia: IArray[A]) apply[A] (i: Int): A = (ia: Array[A])(i)
+
+    // return a sorted copy of the array
+    def sorted[A <: AnyRef : math.Ordering](ia: IArray[A]): IArray[A] = {
```

Comment:
Might be a cool example to also define SortedIArray and return that here. Then binarySearch takes a SortedIArray.